File: app/src/main/java/com/giaquino/android/cropped/ui/base/BaseAuthFragment.java
package com.giaquino.android.cropped.ui.base;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.giaquino.android.cropped.NavGraphMainDirections;
import com.giaquino.android.cropped.common.Constant;
import com.giaquino.android.cropped.data.repository.SharedRepository;

import javax.inject.Inject;

import timber.log.Timber;

public abstract class BaseAuthFragment extends BaseFragment {

    @Inject
    SharedRepository sharedRepository;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setViewVisibility(false);

        if (isLoginCancelled()) {
            onLoginCancelled();
            return;
        }

        if (sharedRepository.isLoggedIn()) {
            onLoginCompleted();
        } else {
            startLogin();
        }
    }

    /**
     * saved state handle will have value once login is cancelled from LoginOauthFragment
     */
    private boolean isLoginCancelled() {
        Boolean completed = getCurrentSavedStateHandle().get(Constant.LOGIN_COMPLETED);
        return completed != null && !completed;
    }

    private void startLogin() {
        Timber.d("start login");
        requireNavController().navigate(NavGraphMainDirections.Companion.actionGlobalLogin());
    }

    private void setViewVisibility(boolean visible) {
        getView().setVisibility(visible ? View.VISIBLE : View.GONE);
    }

    protected void onLoginCompleted() {
        Timber.d("login completed");
        setViewVisibility(true);
    }

    protected void onLoginCancelled() {
        Timber.d("login cancelled");
        getActivity().finish();
    }
}

File: app/src/main/java/com/giaquino/android/cropped/ui/base/BaseAuthFragment.kt
package com.giaquino.android.cropped.ui.base

import android.os.Bundle
import android.view.View
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.lifecycle.SavedStateHandle
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI
import com.giaquino.android.cropped.NavGraphMainDirections
import com.giaquino.android.cropped.R
import com.giaquino.android.cropped.common.Constant
import com.giaquino.android.cropped.data.repository.SharedRepository
import timber.log.Timber
import javax.inject.Inject

/**
 * Base fragment for handling authentication
 */
abstract class BaseAuthFragment : Fragment(), FragmentInterface {

    @Inject
    lateinit var sharedRepository: SharedRepository

    private val navController by lazy {
        NavHostFragment.findNavController(this)
    }

    override fun requireNavController(): NavController {
        return navController
    }

    override fun getCurrentSavedStateHandle(): SavedStateHandle {
        return requireNavController().currentBackStackEntry!!.savedStateHandle
    }

    override fun getPreviousSavedStateHandle(): SavedStateHandle {
        return requireNavController().previousBackStackEntry!!.savedStateHandle
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        when {
            sharedRepository.getUnsplashToken() != null -> {
                Timber.d("login successful")
                setContentVisibility(true)
                initialize()
                initializeToolbar()
            }
            getCurrentSavedStateHandle().get<Boolean>(Constant.LOGIN_CANCELLED) == true -> {
                Timber.d("login cancelled")
                requireNavController().popBackStack()
            }
            else -> {
                Timber.d("start login")
                setContentVisibility(false)
                requireNavController().navigate(NavGraphMainDirections.actionGlobalNavGraphLogin())
            }
        }
    }

    private fun setContentVisibility(visible: Boolean) {
        requireView().visibility = if (visible) View.VISIBLE else View.GONE
    }

    private fun initializeToolbar() {
        val toolbar: Toolbar = requireView().findViewById(R.id.toolbar) ?: return
        NavigationUI.setupWithNavController(
                toolbar,
                requireNavController(),
                AppBarConfiguration.Builder(requireNavController().graph).build())
    }
}